ark/jsonschema/object.ts

```diff
@@ -0,0 +1,265 @@
+import {
+	rootSchema,
+	type Intersection,
+	type Predicate,
+	type Traversal
+} from "@ark/schema"
+import { getDuplicatesOf, printable } from "@ark/util"
+import type { JsonSchema, Out, Type } from "arktype"
+
+import { parseJsonSchema } from "./json.ts"
+import { JsonSchemaScope } from "./scope.ts"
+
+const parseMinMaxProperties = (
+	jsonSchema: JsonSchema.Object,
+	ctx: Traversal
+) => {
+	const predicates: Predicate.Schema[] = []
+	if ("maxProperties" in jsonSchema) {
+		const maxProperties = jsonSchema.maxProperties
+
+		if ((jsonSchema.required?.length ?? 0) > maxProperties) {
+			ctx.reject({
+				expected: `an object JSON Schema with at most ${jsonSchema.maxProperties} required properties`,
+				actual: `an object JSON Schema with ${jsonSchema.required!.length} required properties`
+			})
+		}
+
+		const jsonSchemaObjectMaxPropertiesValidator = (
+			data: object,
+			ctx: Traversal
+		) => {
+			const keys = Object.keys(data)
+			return keys.length <= maxProperties ?
+					true
+				:	ctx.reject({
+						expected: `an object with at most ${maxProperties} propert${maxProperties === 1 ? "y" : "ies"}`,
+						actual: `an object with ${keys.length.toString()} propert${maxProperties === 1 ? "y" : "ies"}`
+					})
+		}
+		predicates.push(jsonSchemaObjectMaxPropertiesValidator)
+	}
+	if ("minProperties" in jsonSchema) {
+		const minProperties = jsonSchema.minProperties
+
+		const jsonSchemaObjectMinPropertiesValidator = (
+			data: object,
+			ctx: Traversal
+		) => {
+			const keys = Object.keys(data)
+			return keys.length >= minProperties ?
+					true
+				:	ctx.reject({
+						expected: `an object with at least ${minProperties} propert${minProperties === 1 ? "y" : "ies"}`,
+						actual: `an object with ${keys.length.toString()} propert${minProperties === 1 ? "y" : "ies"}`
+					})
+		}
+		predicates.push(jsonSchemaObjectMinPropertiesValidator)
+	}
+	return predicates
+}
+
+const parsePatternProperties = (
+	jsonSchema: JsonSchema.Object,
+	ctx: Traversal
+) => {
+	if (!("patternProperties" in jsonSchema)) return
+
+	const patternProperties = Object.entries(jsonSchema.patternProperties).map(
+		([key, value]) => [new RegExp(key), parseJsonSchema(value)] as const
+	)
+
+	// Ensure that the schema for any property is compatible with any corresponding patternProperties
+	patternProperties.forEach(([pattern, parsedPatternPropertySchema]) => {
```

Comment:
This is great but if we're going to add it, it should be to the ArkType implementation so that everyone benefits from it, not just at the JSON schema level.

This is one of the few cases I know we should be checking to fail fast but currently aren't. Would love if you'd add this logic to `structure.ts` and add some tests like the following to `ark/type/__tests__/props`:

```ts
const t = type({
	"[string]": "number",
	foo: "string"
})

// should be a parse error
```

Reply:
Implemented between 69adbe44 and 43e2c304.

Per discord conversations also updated ArkErrors intersection logic to have `never` errors override any other errors, rather than merging them: d000e13c

Thanks for your assistance with implementing this! :)